src/FromData.purs

```diff
@@ -1,31 +1,142 @@
+
+-- | FromDataWithIndex instances for Data.Generic.Rep
+-- See https://purescript-simple-json.readthedocs.io/en/latest/generics-rep.html
+
+instance (FromDataWithIndex l a, FromDataWithIndex r a) => FromDataWithIndex (G.Sum l r) a where
+  fromDataWithIndex _ pci pd = G.Inl <$> fromDataWithIndex (Proxy :: Proxy l) pci pd
+    <|> G.Inr <$> fromDataWithIndex (Proxy :: Proxy r) pci pd
+
+instance (IsSymbol n, FromDataArgs arg) => FromDataWithIndex (G.Constructor n arg) a where
+  fromDataWithIndex _ pci (Constr i pdArgs) = do
+    ix <- BigInt.toInt i
+    cn <- resolveConstr pci ix
+    let rn = reflectSymbol (Proxy :: Proxy n)
+    guard $ cn == rn
+    { head: repArgs, tail: pdArgs' } <- fromDataArgs pdArgs
+    guard $ pdArgs' == []
+    pure $ G.Constructor repArgs
+  fromDataWithIndex _ _ _ = Nothing
+
+instance (HasConstrIndex ci, FromDataWithIndex a ci) => FromDataWithIndex (G.Argument a) ci where
+  fromDataWithIndex _ pci pd = G.Argument <$> fromDataWithIndex (Proxy :: Proxy a) pci pd
+
+-- | FromDataArgs instance for Data.Generic.Rep
+
+instance FromDataArgs (G.NoArguments) where
+  fromDataArgs [] = Just { head: G.NoArguments, tail: [] }
+  fromDataArgs _ = Nothing -- TODO: Error "No PlutusData expected but got some"
+
+instance (FromDataArgsRL list row, RL.RowToList row list) => FromDataArgs (G.Argument (Record row)) where
+  fromDataArgs pdArgs = do
+    { head, tail } <- fromDataArgsRec (Proxy :: Proxy list) pdArgs
+    pure { head: G.Argument head, tail }
+else instance (FromData a) => FromDataArgs (G.Argument a) where
+  fromDataArgs pdArgs = do
+    { head: pd, tail: pds } <- uncons pdArgs -- TODO: Error "Expected PlutusData but got none"
+    repArg <- fromData pd
+    pure $ { head: G.Argument repArg, tail: pds }
+
+instance (FromDataArgs a, FromDataArgs b) => FromDataArgs (G.Product a b) where
+  fromDataArgs pdArgs = do
+    { head: repFst, tail: pdArgs' } <- fromDataArgs pdArgs
+    { head: repSnd, tail: pdArgs'' } <- fromDataArgs pdArgs'
+    pure $ { head: G.Product repFst repSnd, tail: pdArgs'' }
+
+-- | FromDataArgsRL instances
+
+instance FromDataArgsRL RL.Nil () where
+  fromDataArgsRec _ [] = Just { head: {}, tail: [] }
+  fromDataArgsRec _ _ = Nothing
+
+instance
+  ( FromData a
+  , FromDataArgsRL listRest rowRest
+  , Row.Lacks key rowRest
+  , Row.Cons key a rowRest rowFull
+  , RL.RowToList rowFull (RL.Cons key a listRest)
+  , IsSymbol key
+  ) =>
+  FromDataArgsRL (RL.Cons key a listRest) rowFull where
+  fromDataArgsRec _ pdArgs = do
+    let keyProxy = Proxy :: Proxy key
+    { head: pdArg, tail: pdArgs' } <- uncons pdArgs -- TODO: Error "Expected PlutusData but got none"
+    field <- fromData pdArg
+    { head: rec, tail: pdArgs'' } <- fromDataArgsRec (Proxy :: Proxy listRest) pdArgs'
+    pure $
+      { head: (Record.insert keyProxy field rec)
+      , tail: pdArgs''
+      }
+
+genericFromData :: forall a rep. G.Generic a rep => HasConstrIndex a => FromDataWithIndex rep a => PlutusData -> Maybe a
```

Comment:
Our unofficial PS style guide (which we should really write down) recommends kind signatures on all type variables (even when the kind can be completely inferred)